torch/_refs/linalg/__init__.py

```diff
@@ -0,0 +1,103 @@
+):
+    # Checks
+    linalg.utils.check_fp_or_complex(x, "linalg.vector_norm", half=True)
+
+    if isinstance(dim, int):
+        dim = [dim]  # type: ignore[assignment]
+    elif not isinstance(dim, List) and dim is not None:
+        # refs.sum just accepts List rather than DimType
+        dim = list(dim)  # type: ignore[assignment]
+
+    # TODO These things are TORCH_CHECKS, replace them with check(a,b)
+    # once https://github.com/pytorch/pytorch/pull/78014 is merged
+    if x.numel() == 0 and (ord < 0.0 or ord == float("inf")):
+        if dim is None or len(dim) == 0:
+            raise RuntimeError(
+                "linalg.vector_norm cannot compute the {ord} norm on an empty tensor "
+                "because the operation does not have an identity"
+            )
+        else:
+            for d in dim:
+                if x.size(d) == 0:
+                    raise RuntimeError(
+                        f"linalg.vector_norm cannot compute the {ord} norm on the "
+                        "dimension {d} because this dimension is empty and the "
+                        "operation does not have an identity"
+                    )
+    linalg.utils.check_norm_dtype(dtype, x.dtype, "linalg.vector_norm")
+
+    computation_dtype, result_dtype = reduction_dtypes(
+        x, REDUCTION_OUTPUT_TYPE_KIND.COMPLEX_TO_FLOAT, dtype
+    )
+
+    to_result_dtype = partial(prims.convert_element_type, dtype=result_dtype)
+
+    # Implementation
+    if ord == 0.0:
+        return refs.sum(refs.ne(x, 0.0), dim=dim, keepdim=keepdim, dtype=result_dtype)
+    elif ord == float("inf"):
+        return to_result_dtype(refs.amax(prims.abs(x), dim=dim, keepdim=keepdim))
+    elif ord == float("-inf"):
+        return to_result_dtype(refs.amin(prims.abs(x), dim=dim, keepdim=keepdim))
+    else:
+        # From here on the computation dtype is important as the reduction is non-trivial
+        x = prims.convert_element_type(x, computation_dtype)
+        reduce_sum = partial(refs.sum, dim=dim, keepdim=keepdim)
+
+        def fast_pow(x, p):
+            if p == 1.0:
+                return x
+            elif p == 2.0:
+                return prims.mul(x, x)
+            elif p == 0.5:
+                return prims.sqrt(x)
+            else:
+                return prims.pow(x, p)
+
+        # Avoid computing a sqrt in abs and then squaring (more stable)
+        # This could potentially be done for complex dtypes as
+        # x = prims.real(prims.mul(prims.conj(x), x))
+        # and it should be more stable, but it's not clear whether it'll be faster on, say
+        # CPU (abs is 1 vectorised operation), so leaving it just for real dtypes for now
+        ord_pow = ord
+        if ord % 2.0 == 0.0 and is_float_dtype(x.dtype):
+            x = prims.mul(x, x)
+            ord_pow /= 2.0
+        else:
+            x = prims.abs(x)
+        return to_result_dtype(fast_pow(reduce_sum(fast_pow(x, ord_pow)), 1.0 / ord))
```

Comment:
I'm not sure this captures `out` behavior for norms that requires `out` to be of the same type as `dtype` if specified.

Reply:
I would assume so, as the result is of dtype `dtype`, and the out wrapper would use it to check against `out`. Now, I'll let @mruberry confirm

Reply:
`out` wrapper copies to any `out` that's of a weakly higher type (e.g. `float` result can be copied to `double` or `cfloat` `out`)

Reply:
I think it's also OK for the reference to be divergent from the torch implementation here, right? Like ideally wouldn't the torch reference perform the safe copy, too?

Reply:
Idk, other reduction references are not divergent.

Reply:
Whatever's easiest

Reply:
I did not know that we allowed that `out` behaviour. I thought that `out` needed to be of the exact right type. If that's not the case, the `out` behaviour for very many operations in PyTorch is wrong. Also, afaik, this is not tested in the OpInfos. I think if we wanted to enforce this behaviour we should start by adding a test to the `test_out` in the OpInfos.

Reply:
I think I'm missing something here then. `linalg.vector_norm`'s OpInfo does not have any skip, so it should abide by everything we're testing. As such, the out wrapper should behave as expected, as it matches what the `OpInfo` is testing?

Reply:
We test that out doesn't allow unsafe casting, but we don't test that it supports safe casting currently, because our support for safe casting is spotty and I didn't want to skip the entire test. We should just add a different test for it (like the out warning test).

Reply:
See https://github.com/pytorch/pytorch/blob/d990277908a9bf57ddeffe774842544d9920359e/test/test_ops.py#L877

Reply:
I understand now, thank you!

So, at the moment, `vector_norm` is a structured kernel, so it supports as many (or as few) features as all the functions that are implemented as structured kernels. So, if this is a problem with structured kernels in general, it should be solved at the level of structured kernels. If it is not... well, then there's nothing to solve :)

Reply:
We don't need to fix it in the core, but references should be faithfully reproducing this behavior.